python_onnxifi/python_onnxifi.cc

```diff
@@ -0,0 +1,258 @@
+      deviceString = deviceHandle.substr(0, position);
+    }
+    auto it = string_to_deviceType_.find(deviceString);
+    if (it == string_to_deviceType_.end())  {
+      return NULL;
+    }
+    auto deviceType = it->second;
+    //Treating ID of 0 as 1 and bringing 1-indexing to 0-indexing
+    if (deviceID > 0) {
+      deviceID--;
+    }
+    if (deviceID < 0 || deviceID >= device_to_onnxBackendID_[deviceType].size())  {
+      return NULL;
+    }
+    return device_to_onnxBackendID_[deviceType][deviceID];
+  }
+
+  //Input is a string device handle 
+    //(e.g. "CPU" or "GPU:1" or "CPU")
+    //ID of 0 can mean any device of the given type (default to ID 1 for now)
+  //Initializes backend if not initialized
+  //Updates initialized_ map
+  //Returns onnxBackend pointer (NULL if input string is not a valid device)
+  onnxBackend prepDevice(const std::string& deviceHandle)  {
+    auto id = this->getBackendID(deviceHandle);
+    if (!id)  {
+      return NULL;
+    }
+    if (initialized_.find(id) == initialized_.end())  {
+      onnxBackend backend;
+      if (onnxInitBackend(id, NULL, &backend) != ONNXIFI_STATUS_SUCCESS)  {
+        throw("Internal error: onnxInitBackend failed (expected ONNXIFI_STATUS_SUCCESS)");
+      }
+      initialized_[id] = backend;
+    }
+    return initialized_[id];
+  }
+
+  //Release initialized backends
+  //Release all backendIDs
+  ~DeviceIDs()  {
+    for (auto backend : initialized_) {
+      if (onnxReleaseBackend(backend.second) != ONNXIFI_STATUS_SUCCESS)  {
+        throw("Internal error: onnxReleaseBackend failed (expected ONNXIFI_STATUS_SUCCESS)");
+      }
+    }
+    if (!ids_) {
+      for (auto i = 0; i < num_backends_; ++i)  {
+        if (onnxReleaseBackendID(ids_[i]) != ONNXIFI_STATUS_SUCCESS)  {
+          throw("Internal error: onnxReleaseBackendID failed (expected ONNXIFI_STATUS_SUCCESS)");
+        }
+      }
+      delete [] ids_;
+    }
+  }
+
+  //Convenience maps to map betweeen onnxEnum and string
+  //  python user will input the string
+  //  C++ internals and ONNXIFI represent as onnxEnum
+  static const std::unordered_map<onnxEnum, std::string> deviceType_to_string_; 
+  static const std::unordered_map<std::string, onnxEnum> string_to_deviceType_; 
+
+private: 
+  //Array of all backendIDs
+  onnxBackendID* ids_;
+  size_t num_backends_;
+
+  //device_to_onnxBackendID[deviceType][deviceID - 1] will give corresponding onnxBackendID
+  //1-indexed to conform to onnx/onnx/backend/base.py
+  //deviceID of 0 refers to any onnxBackendID of that deviceType
+  std::unordered_map<onnxEnum, std::vector<onnxBackendID>> device_to_onnxBackendID_;
+
+  //If the backendID has been initialized, will map to corresponding backendID pointer
+  //Otherwise, not present
+  std::unordered_map<onnxBackendID, onnxBackend> initialized_;
+};
+
+const std::unordered_map<onnxEnum, std::string> DeviceIDs::deviceType_to_string_ = {{ONNXIFI_DEVICE_TYPE_NPU, "NPU"},
```

Comment:
this can cause SIOF issue https://isocpp.org/wiki/faq/ctors#static-init-order, need to wrap it into a function.

Reply:
didn't think about potential static dependencies here - fix add in separate PR